### PR TITLE
Allow shapes to be constructed with zero values

### LINKED ISSUE #13332: `RegularPolygon` should allow `circumradius` to be 0.0
## What problem does this solve or what need does it fill?
It currently panics when `circumradius` is zero.

https://github.com/bevyengine/bevy/blob/443ce9a62b6bb0136c192a6fe46e4c254a895e6c/crates/bevy_math/src/primitives/dim2.rs#L729

Is there any advantage to rejecting 0.0?

Since the value 0.0 is frequently used as a default or minimum value, not allowing it would increase the chances of the game panicking.

Consider, for example, an animation that smoothly resizes a circle. When the circle appears, the radius increases from zero, and when the circle disappears, the radius decreases toward zero. Implementing this animation in v0.12 is simple, but v0.13 requires additional code to make the radius greater than zero to prevent panic.

## What solution would you like?
That assertion should be `circumradius >= 0.0` to allow zero.

## Additional context
Bevy v0.12 can handle a circle of radius 0.0, but v0.13 panics.

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, startup)
        .run();
}

fn startup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<ColorMaterial>>,
) {
    commands.spawn(Camera2dBundle::default());
    commands.spawn((
        // v0.12 (no panic)
        // ColorMesh2dBundle {
        //     mesh: meshes.add(shape::Circle::new(0.0).into()).into(),
        //     material: materials.add(Color::WHITE.into()),
        //     ..default()
        // },

        // v0.13 (panics)
        ColorMesh2dBundle {
            mesh: meshes.add(Circle::new(0.0)).into(),
            material: materials.add(Color::WHITE),
            ..default()
        },
    ));
}
```

### PR DESCRIPTION
**Pull request by juliohq**
# Objective

Fixes #13332.

## Solution

The assertion `circumradius >= 0.0` to allow zero.

Are there any other shapes that need to be allowed to be constructed with zero?

### CODE REVIEW

**Comment by mamekoro** [comment]
Thanks for the pull request. I opened that issue, but it took me a while to investigate.

It should be sufficient to fix `RegularPolygon`.

I tested the following 2D and 3D shapes implementing `Into<Mesh>` and `Primitive{2,3}d`.

- 2D shapes
    - Annulus
    - Capsule2d
    - Circle
    - Ellipse
    - Rectangle
    - RegularPolygon
    - Triangle2d
- 3D shapes
    - Capsule3d
    - Cone
    - Cuboid
    - Cylinder
    - Plane3d
    - Sphere
    - Torus
    - Triangle3d

Only `Circle` and `RegularPolygon` cause a panic when initialized with zero values.
Circle meshes are instantiated using `RegularPolygon` (in `CircleMeshBuilder`), so fixing `RegularPolygon` should also fix `Circle`.

**Comment by juliohq** [comment]
@mamekoro I couldn't get `Circle` to panic. Could you share the test you've used?

**Comment by mamekoro** [comment]
```sh
cargo new circle
cd circle
git clone --depth=1 https://github.com/bevyengine/bevy
cargo add --path bevy
```

`./src/main.rs` is this;
```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, startup)
        .run();
}

fn startup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<ColorMaterial>>,
) {
    commands.spawn(Camera2dBundle::default());
    commands.spawn(ColorMesh2dBundle {
        mesh: meshes.add(Circle::new(0.0)).into(),
        material: materials.add(Color::WHITE),
        ..default()
    });
}
```

<details>
<summary>full test</summary>

It is difficult to tell which shape has panicked when testing multiple shapes at once, so toggle comments to select only one shape to test.

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, test2d)
        //.add_systems(Startup, test3d)
        .run();
}

fn test2d(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<ColorMaterial>>,
) {
    commands.spawn(Camera2dBundle::default());

    // panics
    commands.spawn(ColorMesh2dBundle {
        mesh: meshes.add(Circle::new(0.0)).into(),
        material: materials.add(Color::WHITE),
        ..default()
    });

    // no panic
    //commands.spawn(ColorMesh2dBundle {
    //    mesh: meshes.add(Ellipse::new(0.0, 0.0)).into(),
    //    material: materials.add(Color::WHITE),
    //    ..default()
    //});

    // no panic
    //commands.spawn(ColorMesh2dBundle {
    //    mesh: meshes.add(Annulus::new(0.0, 0.0)).into(),
    //    material: materials.add(Color::WHITE),
    //    ..default()
    //});

    // no panic
    //commands.spawn(ColorMesh2dBundle {
    //    mesh: meshes
    //        .add(Triangle2d::new(Vec2::ZERO, Vec2::ZERO, Vec2::ZERO))
    //        .into(),
    //    material: materials.add(Color::WHITE),
    //    ..default()
    //});

    // no panic
    //commands.spawn(ColorMesh2dBundle {
    //    mesh: meshes.add(Rectangle::new(0.0, 0.0)).into(),
    //    material: materials.add(Color::WHITE),
    //    ..default()
    //});

    // panics
    //commands.spawn(ColorMesh2dBundle {
    //    mesh: meshes.add(RegularPolygon::new(0.0, 3)).into(),
    //    material: materials.add(Color::WHITE),
    //    ..default()
    //});

    // no panic
    //commands.spawn(ColorMesh2dBundle {
    //    mesh: meshes.add(Capsule2d::new(0.0, 0.0)).into(),
    //    material: materials.add(Color::WHITE),
    //    ..default()
    //});
}

fn test3d(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    commands.spawn(Camera3dBundle {
        transform: Transform::from_xyz(0.0, 5.0, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
        ..default()
    });
    commands.spawn(PointLightBundle {
        point_light: PointLight {
            shadows_enabled: true,
            intensity: 10_000_000.,
            range: 100.0,
            ..default()
        },
        transform: Transform::from_xyz(5.0, 10.0, 5.0),
        ..default()
    });

    // no panic
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Capsule3d::new(0.0, 0.0)),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});

    // no panic
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Cone {
    //        radius: 0.0,
    //        height: 0.0,
    //    }),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});

    // no panic
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Cuboid {
    //        half_size: Vec3::ZERO,
    //    }),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});

    // no panic
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Cylinder::new(0.0, 0.0)),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});

    // no panic (Tested with a non-zero normal vector)
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Plane3d::new(Vec3::new(0.0, 1.0, 0.0), Vec2::ZERO)),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});

    // no panic
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Sphere::new(0.0)),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Sphere::new(0.0).mesh().ico(4).unwrap()),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Sphere::new(0.0).mesh().uv(4, 4)),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});

    // no panic
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Torus::new(0.0, 0.0)),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});

    // no panic
    //commands.spawn(PbrBundle {
    //    mesh: meshes.add(Triangle3d::new(Vec3::ZERO, Vec3::ZERO, Vec3::ZERO)),
    //    material: materials.add(StandardMaterial::default()).clone(),
    //    ..default()
    //});
}
```

</details>

**Comment by juliohq** [comment]
@mamekoro I just ran your test against my changes and they seem to solve the issue. Thank you.

File: crates/bevy_math/src/primitives/dim2.rs [hunk 1]
```diff
@@ -723,10 +723,10 @@ impl RegularPolygon {
     ///
     /// # Panics
     ///
-    /// Panics if `circumradius` is non-positive
+    /// Panics if `circumradius` is negative
     #[inline(always)]
     pub fn new(circumradius: f32, sides: usize) -> Self {
-        assert!(circumradius > 0.0, "polygon has a non-positive radius");
+        assert!(circumradius >= 0.0, "polygon has a negative radius");
```
Comment (mockersf) [inline crates/bevy_math/src/primitives/dim2.rs:729]:
```suggestion
        assert!(circumradius.is_sign_positive(), "polygon has a negative radius");
```

**Review by ramirezmike — COMMENTED** [review]
Looks good to me, although I agree with [mockersf](https://github.com/mockersf)'s suggestion

**Comment by mamekoro** [comment]
Should we reject `-0.0`?
`assert!(circumradius.is_sign_positive())` panics when `circumradius` is `-0.0`, but `assert!(circumradius >= 0.0)` does not.

**Comment by alice-i-cecile** [comment]
> Should we reject `-0.0`? `assert!(circumradius.is_sign_positive())` panics when `circumradius` is `-0.0`, but `assert!(circumradius >= 0.0)` does not. Bevy v0.12 also doesn't panic on radius `-0.0`.

I would allow it: it should behave fine.

**Review by alice-i-cecile — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

**Pull request merged**